### PR TITLE
Filemanager - bulk download, bulk delete

### PR DESCRIPTION
**Pull request by tenisakb**

### CODE REVIEW

File: src/components/file-manager/file-manager-slide.tsx [hunk 1]
```diff
@@ -146,30 +152,37 @@ const Table = styled.table`
...
+const Tr = styled.tr<TrProps>`
   &:hover td {
     background-color: #cfe3fc;
   }
-  th:first-child,
-  td:first-child {
+  th:nth-child(1),
+  td:nth-child(1),
+  th:nth-child(${props => props.move + 1}),
+  td:nth-child(${props => props.move + 1}) {
```
[the remark this answers is not in the document]
Comment (EdmundMai) [reply src/components/file-manager/file-manager-slide.tsx]: i think you can style this without this nth child stuff

**Comment by funkydrummer** [comment]
image


Example of what the dropdown would look like

File: src/components/file-manager/file-manager-slide.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (tenisakb) [reply src/components/file-manager/file-manager-slide.tsx]: done

File: src/redux/actions/filemanager-actions.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  resetFiles: () => ({
+    type: RESET_FILES
+  }),
+  setFiles: ({ files }) => ({
+    type: SET_FILES,
+    payload: { files }
+  })
+});
+
+export default ACTIONS;
```
[the remark this answers is not in the document]
Comment (tenisakb) [reply src/redux/actions/filemanager-actions.ts]: done

File: src/components/file-manager/file-manager-slide.tsx [hunk 3]
```diff
@@ -341,14 +380,77 @@ const FileManagerSlide = ({
...
 
-  useEffect(
-    () => {
-      const defaultOrder = "created";
-      setOrderedFiles(_.orderBy(files, defaultOrder, "desc"));
-      setParam(defaultOrder);
-    },
-    [files]
-  );
+  const setFilemanagerFile = (handle: string) => {
```
Comment (EdmundMai) [inline src/components/file-manager/file-manager-slide.tsx]: rename to `selectFile`

File: src/components/file-manager/file-manager-slide.tsx [hunk 4]
```diff
@@ -341,14 +380,77 @@ const FileManagerSlide = ({
...
+    const files: Handle[] = [...filemanagerFiles, { handle }];
+    setFilemanagerFiles(files);
+  };
+
+  const deleteFilemanagerFile = (handle: string) => {
```
Comment (EdmundMai) [inline src/components/file-manager/file-manager-slide.tsx]: rename to `deselectFile`

File: src/components/file-manager/file-manager-slide.tsx [hunk 5]
```diff
@@ -341,14 +380,77 @@ const FileManagerSlide = ({
...
-    () => {
-      const defaultOrder = "created";
-      setOrderedFiles(_.orderBy(files, defaultOrder, "desc"));
-      setParam(defaultOrder);
-    },
-    [files]
-  );
+  const setFilemanagerFile = (handle: string) => {
+    const files: Handle[] = [...filemanagerFiles, { handle }];
+    setFilemanagerFiles(files);
```
Comment (EdmundMai) [inline src/components/file-manager/file-manager-slide.tsx]: just combine 384-385 to: `setFilemanagerFiles([...filemanagerFiles, { handle }])`

File: src/components/file-manager/file-manager-slide.tsx [hunk 6]
```diff
@@ -341,14 +380,77 @@ const FileManagerSlide = ({
...
+  };
+
+  const deleteFilemanagerFile = (handle: string) => {
+    setFilemanagerFiles(
+      filemanagerFiles.filter(file => file.handle !== handle)
+    );
+  };
+
+  const checkedAll = () => {
```
Comment (EdmundMai) [inline src/components/file-manager/file-manager-slide.tsx]: rename to `checkedAllIcon`

File: src/components/file-manager/file-manager-slide.tsx [hunk 7]
```diff
@@ -341,14 +380,77 @@ const FileManagerSlide = ({
...
+      />
+    );
+  };
+
+  const PickFile = ({ handle }) => {
+    if (orderedFiles.length === filemanagerFiles.length) {
+      setCheckedAllState(true);
+    } else {
+      setCheckedAllState(false);
+    }
```
Comment (EdmundMai) [inline src/components/file-manager/file-manager-slide.tsx]: this doesn't belong here, do it when you are selecting / deselecting a file, not when you're loading an icon

File: src/components/file-manager/file-manager-slide.tsx [hunk 8]
```diff
@@ -341,14 +380,77 @@ const FileManagerSlide = ({
...
+      />
+    ) : (
+      <TableIcon
+        onClick={() => [setCheckedAllState(false), setFilemanagerFiles([])]}
+        src={ICON_CHECKBOX}
+      />
+    );
+  };
+
+  const PickFile = ({ handle }) => {
```
Comment (EdmundMai) [inline src/components/file-manager/file-manager-slide.tsx]: rename to `checkFileIcon`

File: src/components/forgot-page/forgot-page-slide.tsx [hunk 9]
```diff
@@ -64,8 +64,49 @@ const Label = styled.h3`
...
+        letter !== "," &&
+        letter !== ";" &&
+        letter !== "\t" &&
+        letter !== "&nbsp;"
+      ) {
+        result += letter;
+      } else {
+        result += " ";
+      }
+    });
```
Comment (EdmundMai) [inline src/components/forgot-page/forgot-page-slide.tsx]: use `Array.reduce` instead

`value.split("").reduce( ... )`

File: src/config/index.ts [hunk 10]
```diff
@@ -261,3 +261,9 @@ export const DATA_TYPES_ICONS = [
     icon: ICON_PNG
   }
 ];
+
+export const MULTIPLE_ACTIONS = {
```
Comment (EdmundMai) [inline src/config/index.ts]: use `enum`. we use it in this file

```
export enum MULTIPLE_ACTIONS {
  MULTIPLE_REMOVE,
  ...
}
```

Comment (tenisakb) [reply src/config/index.ts]: done

File: src/components/forgot-page/forgot-page-slide.tsx (hunk 9 above)
Comment (tenisakb) [reply src/components/forgot-page/forgot-page-slide.tsx]: done

File: src/components/file-manager/file-manager-slide.tsx (hunk 8 above)
Comment (tenisakb) [reply src/components/file-manager/file-manager-slide.tsx]: done

File: src/components/file-manager/file-manager-slide.tsx (hunk 7 above)
Comment (tenisakb) [reply src/components/file-manager/file-manager-slide.tsx]: done

File: src/components/file-manager/file-manager-slide.tsx (hunk 6 above)
Comment (tenisakb) [reply src/components/file-manager/file-manager-slide.tsx]: done

File: src/components/file-manager/file-manager-slide.tsx (hunk 5 above)
Comment (tenisakb) [reply src/components/file-manager/file-manager-slide.tsx]: done

File: src/components/file-manager/file-manager-slide.tsx (hunk 4 above)
Comment (tenisakb) [reply src/components/file-manager/file-manager-slide.tsx]: done

File: src/config/index.ts [hunk 11]
```diff
@@ -261,3 +261,8 @@ export const DATA_TYPES_ICONS = [
     icon: ICON_PNG
   }
 ];
+
+export enum MULTIPLE_ACTIONS {
+  MULTIPLE_REMOVE,
+  MULTIPLE_DOWNLOAD
+}
```
Comment (EdmundMai) [inline src/config/index.ts]: name this to:

```
export enum BULK_ACTIONS {
  REMOVE,
  DOWNLOAD
}```

File: src/components/file-manager/file-manager-slide.tsx [hunk 12]
```diff
@@ -388,9 +478,34 @@ const FileManagerSlide = ({
                   onSelected={files => upload(files, masterHandle)}
                 />
               </ButtonWrapper>
+              <MultiActionContainer>
+                <BulkSelect
+                  value={multipleAction}
+                  onChange={event =>
+                    setMultipleAction(
```
Comment (EdmundMai) [inline src/components/file-manager/file-manager-slide.tsx]: rename to `setBulkAction`

File: src/redux/actions/authentication-actions.test.js [hunk 13]
```diff
@@ -135,3 +135,10 @@ test("resetAccountHandle", () => {
   };
   expect(actions.resetAccountHandle()).toEqual(expected);
 });
+
+test("resetAccountHandle", () => {
```
Comment (EdmundMai) [inline src/redux/actions/authentication-actions.test.js]: `resetRecoverError`

File: src/redux/reducers/authentication-reducer.test.js [hunk 14]
```diff
@@ -114,11 +116,50 @@ test("authentication-reducer RESET_ACCOUNT_HANDLE", () => {
...
+  const state = {
+    status: AUTHENTICATION_STATUSES.LOGGED_OUT,
+    masterHandle: null,
+    storageUsed: 0,
+    storageLimit: 0,
+    expirationDate: null,
+    recoveryHandle: null,
+    error: true,
+    errorCode: "aa"
+  };
```
Comment (EdmundMai) [inline src/redux/reducers/authentication-reducer.test.js]: i think you can use `initState` here too.
```
const error = new Error("e1");
const state = { ...initState, error };
...
```

File: src/redux/reducers/authentication-reducer.ts [hunk 15]
```diff
@@ -7,7 +7,9 @@ const initState = {
   storageUsed: 0,
   storageLimit: 0,
   expirationDate: null,
-  recoveryHandle: null
+  recoveryHandle: null,
+  error: false,
+  errorCode: ""
```
Comment (EdmundMai) [inline src/redux/reducers/authentication-reducer.ts]: i don't think you're using `errorCode` anywhere, delete it

File: src/redux/reducers/authentication-reducer.ts [hunk 16]
```diff
@@ -7,7 +7,9 @@ const initState = {
   storageUsed: 0,
   storageLimit: 0,
   expirationDate: null,
-  recoveryHandle: null
+  recoveryHandle: null,
+  error: false,
```
Comment (EdmundMai) [inline src/redux/reducers/authentication-reducer.ts]: either rename this to make it look more like a boolean (e.g. `hasError`) or change the value to be either `null || Error`. error sounds like it's supposed to be an Error object

File: src/components/file-manager/file-manager-slide.tsx [hunk 17]
```diff
@@ -341,14 +379,66 @@ const FileManagerSlide = ({
...
+      );
+    }
+    return <TableIcon src={ICON_REMOVE} onClick={() => selectFile(handle)} />;
+  };
+
+  const bulkActions = () => {
+    if (filemanagerFiles.length) {
+      if (multipleAction === MULTIPLE_ACTIONS.MULTIPLE_DOWNLOAD) {
+        downloadFiles(filemanagerFiles);
+      } else {
```
[the remark this answers is not in the document]
Comment (EdmundMai) [reply src/components/file-manager/file-manager-slide.tsx]: or better yet use a switch statement

File: src/redux/reducers/authentication-reducer.ts (hunk 16 above)
Comment (tenisakb) [reply src/redux/reducers/authentication-reducer.ts]: fixed

File: src/redux/reducers/authentication-reducer.ts (hunk 15 above)
Comment (tenisakb) [reply src/redux/reducers/authentication-reducer.ts]: fixed

File: src/redux/reducers/authentication-reducer.test.js (hunk 14 above)
Comment (tenisakb) [reply src/redux/reducers/authentication-reducer.test.js]: fixed

File: src/redux/actions/authentication-actions.test.js (hunk 13 above)
Comment (tenisakb) [reply src/redux/actions/authentication-actions.test.js]: fixed

File: src/components/file-manager/file-manager-slide.tsx (hunk 12 above)
Comment (tenisakb) [reply src/components/file-manager/file-manager-slide.tsx]: fixed

File: src/config/index.ts (hunk 11 above)
Comment (tenisakb) [reply src/config/index.ts]: fixed

File: src/components/file-manager/file-manager-slide.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (tenisakb) [reply src/components/file-manager/file-manager-slide.tsx]: fixed

File: src/components/file-manager/file-manager-slide.tsx [hunk 18]
```diff
@@ -341,14 +379,66 @@ const FileManagerSlide = ({
...
+    ) : (
+      <TableIcon
+        onClick={() => [setFilemanagerFiles([])]}
+        src={ICON_CHECKBOX}
+      />
+    );
+  };
+
+  const CheckFileIcon = ({ handle }) => {
+    if (filemanagerFiles) {
```
[the remark this answers is not in the document]
Comment (tenisakb) [reply src/components/file-manager/file-manager-slide.tsx]: fixed

File: src/components/file-manager/file-manager-slide.tsx [hunk 19]
```diff
@@ -341,14 +379,66 @@ const FileManagerSlide = ({
...
+      return file ? (
+        <TableIcon src={ICON_CHECKBOX} onClick={() => deselectFile(handle)} />
+      ) : (
+        <TableIcon
+          src={ICON_CHECKBOX_EMPTY}
+          onClick={() => selectFile(handle)}
+        />
+      );
+    }
+    return <TableIcon src={ICON_REMOVE} onClick={() => selectFile(handle)} />;
```
[the remark this answers is not in the document]
Comment (tenisakb) [reply src/components/file-manager/file-manager-slide.tsx]: fixed

File: src/components/file-manager/file-manager-slide.tsx [hunk 20]
```diff
@@ -319,11 +351,17 @@ const FileManagerSlide = ({
...
-  isOver
+  isOver,
+  downloadFiles,
+  removeFiles
 }) => {
   const [orderedFiles, setOrderedFiles] = useState<File[]>([]);
   const [param, setParam] = useState("");
   const [sharedFile, setSharedFile] = useState<File | null>(null);
+  const [bulkAction, setBulkAction] = useState<MULTIPLE_ACTIONS>(
+    MULTIPLE_ACTIONS.DOWNLOAD
```
Comment (EdmundMai) [inline src/components/file-manager/file-manager-slide.tsx]: rename to `BULK_ACTIONS`

File: src/components/forgot-page/forgot-page-slide.tsx [hunk 21]
```diff
@@ -84,13 +121,21 @@ const ForgotPageSlide = ({ recoverAccountHandle }) => {
...
+                  value={mnemonic}
+                  ref={recoverTextearea}
+                  onChange={e => validateMnemonic(e.target.value)}
+                  name="storage-pin"
+                />
               </InputColumnWrapper>
             </InputWrapper>
             <ButtonWrapper>
+              {(!isValidMnemonic || hasError) && (
+                <Error>Mnemonic are not valid.</Error>
```
Comment (EdmundMai) [inline src/components/forgot-page/forgot-page-slide.tsx]: Your mnemonic is invalid.

File: src/config/index.ts (hunk 11 above)
Comment (EdmundMai) [reply src/config/index.ts]: looks like it's still there?

File: src/redux/reducers/authentication-reducer.test.js (hunk 14 above)
Comment (EdmundMai) [reply src/redux/reducers/authentication-reducer.test.js]: looks like it's still there?

**Comment by EdmundMai** [comment]
Updated the UI to be checkbox based. I felt the icons were weird

bulk

File: src/components/forgot-page/forgot-page-slide.tsx (hunk 21 above)
Comment (tenisakb) [reply src/components/forgot-page/forgot-page-slide.tsx]: fixed

File: src/components/file-manager/file-manager-slide.tsx (hunk 20 above)
Comment (tenisakb) [reply src/components/file-manager/file-manager-slide.tsx]: fixed

File: src/redux/reducers/authentication-reducer.test.js (hunk 14 above)
Comment (tenisakb) [reply src/redux/reducers/authentication-reducer.test.js]: fixed

File: src/config/index.ts (hunk 11 above)
Comment (tenisakb) [reply src/config/index.ts]: fixed

**Comment by EdmundMai** [comment]
Replaced by 345

**Pull request closed without merging**